checks/evaluation/branch_protection.go

```diff
@@ -60,41 +71,120 @@ const (
 )
 
 // BranchProtection runs Branch-Protection check.
-func BranchProtection(name string, dl checker.DetailLogger,
-	r *checker.BranchProtectionsData,
+func BranchProtection(name string,
+	findings []finding.Finding, dl checker.DetailLogger,
 ) checker.CheckResult {
-	var scores []levelScore
+	expectedProbes := []string{
+		blocksDeleteOnBranches.Probe,
+		blocksForcePushOnBranches.Probe,
+		branchesAreProtected.Probe,
+		branchProtectionAppliesToAdmins.Probe,
+		dismissesStaleReviews.Probe,
+		requiresApproversForPullRequests.Probe,
+		requiresCodeOwnersReview.Probe,
+		requiresLastPushApproval.Probe,
+		requiresUpToDateBranches.Probe,
+		runsStatusChecksBeforeMerging.Probe,
+		requiresPRsToChangeCode.Probe,
+	}
+
+	if !finding.UniqueProbesEqual(findings, expectedProbes) {
+		e := sce.WithMessage(sce.ErrScorecardInternal, "invalid probe results")
+		return checker.CreateRuntimeErrorResult(name, e)
+	}
 
-	// Check protections on all the branches.
-	for i := range r.Branches {
-		var score levelScore
-		b := r.Branches[i]
+	branchScores := make(map[string]*levelScore)
+	warnedBranches := make([]string, 0)
 
+	for i := range findings {
+		f := &findings[i]
+		if f.Outcome == finding.OutcomeNotApplicable {
+			return checker.CreateInconclusiveResult(name, "unable to detect any development/release branches")
+		}
+
+		branchName, err := getBranchName(f)
+		if err != nil {
+			return checker.CreateRuntimeErrorResult(name, err)
+		}
+		if branchName == "" {
+			e := sce.WithMessage(sce.ErrScorecardInternal, "probe is missing branch name")
+			return checker.CreateRuntimeErrorResult(name, e)
+		}
 		// Protected field only indates that the branch matches
 		// one `Branch protection rules`. All settings may be disabled,
 		// so it does not provide any guarantees.
-		protected := !(b.Protected != nil && !*b.Protected)
-		if !protected {
+		protected, err := isBranchProtected(findings, branchName)
+		if err != nil {
+			e := sce.WithMessage(sce.ErrScorecardInternal, err.Error())
+			return checker.CreateRuntimeErrorResult(name, e)
+		}
+
+		if !protected && !contains(warnedBranches, branchName) {
 			dl.Warn(&checker.LogMessage{
-				Text: fmt.Sprintf("branch protection not enabled for branch '%s'", *b.Name),
+				Text: fmt.Sprintf("branch protection not enabled for branch '%s'", branchName),
```

Comment:
question: now that protection status is its own probe, should this logic just be rolled into to the `switch f.Probe` below?  

1. as written it's quadratic on number of branch names (for each finding, search through all findings for protected probe)
2. We probably don't need to maintain `warnedBranches` anymore.

Reply:
I added a loop specifically for this instead of making it part of the main loop.